packages/core/src/lib/object_factory.js

```diff
@@ -1,108 +1,87 @@
 'use strict';
-const patternEngines = require('./pattern_engines');
+
+const _ = require('lodash');
 const path = require('path');
+const patternEngines = require('./pattern_engines');
 
-// patternPrefixMatcher is intended to match the leading maybe-underscore,
+// prefixMatcher is intended to match the leading maybe-underscore,
 // zero or more digits, and maybe-dash at the beginning of a pattern file name we can hack them
 // off and get at the good part.
-const patternPrefixMatcher = /^_?(\d+-)?/;
+const prefixMatcher = /^_?(\d+-)?/;
 
-// Pattern properties
 /**
- * Pattern constructor
+ * Pattern constructor / Pattern properties
+ *
+ * Before changing functionalities of the pattern object please read the following pull requests
+ * to get more details about the behavior of the folder structure
+ * https://github.com/pattern-lab/patternlab-node/pull/992
+ * https://github.com/pattern-lab/patternlab-node/pull/1016
+ *
  * @constructor
  */
 const Pattern = function(relPath, data, patternlab) {
+  this.relPath = path.normalize(relPath); // '00-atoms/00-global/00-colors.mustache'
+
   /**
    * We expect relPath to be the path of the pattern template, relative to the
    * root of the pattern tree. Parse out the path parts and save the useful ones.
    * @param {relPath} relative directory
    * @param {data} The JSON used to render values in the pattern.
    * @param {patternlab} rendered html files for the pattern
    */
-  const pathObj = path.parse(path.normalize(relPath));
-  const info = {};
-  // 00-colors(.mustache) is subbed in 00-atoms-/00-global/00-colors
-  info.hasDir =
-    path.basename(pathObj.dir).replace(patternPrefixMatcher, '') ===
-      pathObj.name.replace(patternPrefixMatcher, '') ||
-    path.basename(pathObj.dir).replace(patternPrefixMatcher, '') ===
-      pathObj.name.split('~')[0].replace(patternPrefixMatcher, '');
-
-  info.dir = info.hasDir ? pathObj.dir.split(path.sep).pop() : '';
-  info.dirLevel = pathObj.dir.split(path.sep).length;
+  const pathObj = path.parse(this.relPath);
+
+  const info = this.getPatternInfo(pathObj, patternlab);
 
-  this.relPath = path.normalize(relPath); // '00-atoms/00-global/00-colors.mustache'
   this.fileName = pathObj.name; // '00-colors'
   this.subdir = pathObj.dir; // '00-atoms/00-global'
-  if ((this.subdir.match(/\w(?=\\)|\w(?=\/)/g) || []).length > 1) {
-    this.subdir = this.subdir.split(/\/|\\/, 2).join(path.sep); // '00-atoms/03-controls/00-button' -> '00-atoms/03-controls'
-  }
   this.fileExtension = pathObj.ext; // '.mustache'
 
-  // this is the unique name, subDir + fileName (sans extension)
-  this.name = '';
-  if (info.hasDir && info.dirLevel > 2) {
-    let variant = '';
-
-    if (this.fileName.indexOf('~') !== -1) {
-      variant = '-' + this.fileName.split('~')[1];
-    }
-    this.name = this.subdir.replace(/[\/\\]/g, '-') + variant;
+  // TODO: Remove if when droping ordering by prefix and keep else code
```

Comment:
Sometimes when we change the ordering by prefix, we can also reduce some code here.